lib/web_ui/lib/src/engine/html/shader_mask.dart

```diff
@@ -0,0 +1,304 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.12
+part of engine;
+
+/// A surface that applies a shader to its children.
+///
+/// Currently there are 2 types of shaders:
+///   - Gradients
+///   - ImageShader
+///
+/// Gradients
+///   The gradients can be applied to the child tree by rendering the gradient
+///   into an image and referencing the image in an svg filter to apply
+///   to DOM tree.
+///
+class PersistedShaderMask extends PersistedContainerSurface
+    implements ui.ShaderMaskEngineLayer {
+  PersistedShaderMask(
+      PersistedShaderMask? oldLayer,
+      this.shader,
+      this.maskRect,
+      this.blendMode,
+      ) : super(oldLayer);
+
+  html.Element? _childContainer;
+  final ui.Shader shader;
+  final ui.Rect maskRect;
+  final ui.BlendMode blendMode;
+  html.Element? _shaderElement;
+  html.Element? _imageElement;
+  bool containerVisible = true;
+
+  @override
+  void adoptElements(PersistedShaderMask oldSurface) {
+    super.adoptElements(oldSurface);
+    _childContainer = oldSurface._childContainer;
+    _shaderElement = oldSurface._shaderElement;
+    oldSurface._childContainer = null;
+    oldSurface._shaderElement = null;
+  }
+
+  @override
+  void discard() {
+    super.discard();
+    // Do not detach the child container from the root. It is permanently
+    // attached. The elements are reused together and are detached from the DOM
+    // together.
+    _childContainer = null;
+  }
+
+  @override
+  html.Element createElement() {
+    html.Element element = defaultCreateElement('flt-shader-mask');
+    html.Element container = html.Element.tag('flt-mask-interior');
```

Comment:
Is the interior element necessary? We typically use it in clipping layers when the clipping element changes the coordinate system. I'm not sure this layer has this problem?

Reply:
Yes otherwise we can't host svg resource.